### PR TITLE
Set timer for temporary index on master only

### PR DESCRIPTION
**Pull request by ashtul**
This PR is related to temporary indexes.
**Issue**
Currently, a timer is set on both master/slave. This caused several issues.
1. Every time an index is used for either ingest or query, its temporary timer is reset. some clusters will pass read calls only to the master and not to the slave. Therefore, we risk the temporary timer might reach time prior to the timer on the master.
2. The slave call rm_call(`DEL`) on all documents. This caused primary replica inconsistency (and crash on CRDT version since the slave should not get the write command directly).

**Solution**
In this PR we set the timer for the index on the master shard only and when an index expires, RM_CALL("FT.DROPINDEX idx DD") is called are erases the index from both master and slave.
In addition, an event of `RedisModuleEvent_ReplicationRoleChanged` will call `RoleChangeCallback`. If the shard role change to `master`, all temporary indexes will register to expire after their full period. If the shard role changed to `slave`, all temporary indexes unregister their timer.

**Limitation**
Re-registering the timer after a failover on the master shard might mean an index takes longer to expire (up to x2 times).

MOD-3365

### CODE REVIEW

**Comment by MeirShpilraien** [comment]
If possible, lets extend the top comment with some more details:
1. What is the current issue (I remeber we talked of a couple of issues)
2. What is the solution
3. What is the limitations with the solution (we said that in worst case we will have index expire after 2 times the actual expiration time)

File: src/notifications.c [hunk 1]
```diff
@@ -374,3 +374,23 @@ void Initialize_RdbNotifications(RedisModuleCtx *ctx) {
...
+    Indexes_SetTempSpecsTimers(TimerOp_Add);
+    break;
+  case REDISMODULE_EVENT_REPLROLECHANGED_NOW_REPLICA:
+    Indexes_SetTempSpecsTimers(TimerOp_Del);
+    break;
+  }
+}
+
+void Initialize_RoleChangeNotifications(RedisModuleCtx *ctx) {
+  if (isFeatureSupported(RM_EVENT_API_SUPPORT)) {
```
Comment (MeirShpilraien) [inline src/notifications.c:391]: The minimum version we support is 6.0.0 so I guess its not needed no?

File: src/spec.c [hunk 2]
```diff
@@ -1272,10 +1290,6 @@ void IndexSpec_LegacyFree(void *spec) {
 
 void IndexSpec_Free(IndexSpec *spec) {
   if (spec->flags & Index_Temporary) {
-    // we are taking the index to a background thread to be released.
-    // before we do it we need to delete it from the index dictionary
-    // to prevent it from been freed again.c
-    dictDelete(specDict_g, spec->name);
```
Comment (MeirShpilraien) [inline src/spec.c:1278]: Why remove it?

**Review by MeirShpilraien — COMMENTED** [review]
IIRC we also discuss on adding `FORCEKEEPDOCS` option and always replicate `FT.DROPINDEX` with this option so we will never delete documents on replica.

File: src/spec.c (hunk 2 above)
Comment (ashtul) [reply src/spec.c:1278]: Since we rm_call `FT.DROPINDEX` and if we have already removed the index from the global dictionary here, we are going to fail.

**Comment by ashtul** [comment]
> IIRC we also discuss on adding `FORCEKEEPDOCS` option and always replicate `FT.DROPINDEX` with this option so we will never delete documents on replica.

We instead send `FT.DROPINDEX` without the `DD` flag.

**Comment by MeirShpilraien** [comment]
> We instead send FT.DROPINDEX without the DD flag.

But the user might delete this index with `FT.DROP` and then it will break no?

**Review by MeirShpilraien — APPROVED** [review]

**Pull request merged**